OpenStack Integration Runner

tserong opened the pull request:
This is the most minimalist implementation I could come up with, to get a feel for how this OpenStack integration thing should look. In its current form, all it does is pool and key creation, and returns some configuration information. Options for specifying what pools to create, supporting multiple OpenStack clusters, and configuring RGW are notably missing.

As mentioned in #986, the idea is that a person, or another tool, could invoke a runner to get the relevant pools and keys created, and extract the information necessary to configure OpenStack to talk to a Ceph cluster. Here's what doing this via `curl` looks like presently. First, you have to login using some shared secret to get an auth token:

````
curl -sSk 192.168.121.177:8000/login \
    -H 'Accept: application/x-yaml' \
    -d username=admin -d sharedsecret=8faffb00-e936-4bfb-912b-ee51c5393eaa \
    -d eauth=sharedsecret
return:
- eauth: sharedsecret
  expire: 1521589316.739898
  perms:
  - .*
  - '@runner'
  - '@wheel'
  start: 1521546116.739897
  token: ff41144485012f7a4a82ea4686579660f252f832
  user: admin
````

Then you can invoke the `openstack.integrate` runner, passing that token in (the name `openstack.integrate` is more or less arbitrary and is of course completely open for discussion):

````
curl -sSk 192.168.121.177:8000/ \
    -H 'Accept: application/x-yaml' \
    -H 'X-Auth-Token: ff41144485012f7a4a82ea4686579660f252f832' \
    -d client=runner \
    -d fun=openstack.integrate
return:
- conf:
    cluster_network: 172.16.2.0/24
    fsid: 41fe92cf-ac2d-3c50-8e96-8344942d2915
    mon_host:
    - 172.16.1.13
    - 172.16.1.11
    - 172.16.1.12
    mon_initial_members:
    - mon3
    - mon1
    - mon2
    public_network: 172.16.1.0/24
  pools:
    backups: {}
    images:
      key: AQCoUapaAAAAABAAwt8H1QzNR4zQ648FENGt+A==
      name: glance
      user: client.glance
    vms:
      key: AQCoUapaAAAAABAAkXn3tFIV+T6HQKOj9MZkmA==
      name: nova
      user: client.nova
    volumes:
      key: AQCoUapaAAAAABAAX9aTpLygI5MmEygAk7gl4w==
      name: cinder
      user: client.cinder
````

I *think* that's enough information in order for a person or tool to subsequently construct appropriate ceph.conf and keyring files for use on the OpenStack side, but would appreciate confirmation.

The "pools" dict above tells the client the user name, secret key and pool name for each of the "volumes", "images", "backups" and "vms" pools suggested at http://docs.ceph.com/docs/master/rbd/rbd-openstack/ (note the the pool names actually created at present are "glance", "cinder" and "nova"; "backups" is missing. I'll change these so that the default pool names are consistent with the usage in the Ceph docs, but probably prefixed with "openstack-", or "openstack-$name-", once I add a parameter to allow multiple pool sets to handle multiple OpenStack clusters sharing one Ceph cluster).

tserong commented:
OK, now the output is as follows:

````
return:
- conf:
    cluster_network: 172.16.2.0/24
    fsid: 41fe92cf-ac2d-3c50-8e96-8344942d2915
    mon_host:
    - 172.16.1.13
    - 172.16.1.11
    - 172.16.1.12
    mon_initial_members:
    - mon3
    - mon1
    - mon2
    public_network: 172.16.1.0/24
  pools:
    cinder:
      key: AQCoUapaAAAAABAAX9aTpLygI5MmEygAk7gl4w==
      rbd_store_pool: cinder
      rbd_store_user: cinder
    cinder-backup: {}
    glance:
      key: AQCoUapaAAAAABAAwt8H1QzNR4zQ648FENGt+A==
      rbd_store_pool: glance
      rbd_store_user: glance
    nova:
      key: AQCoUapaAAAAABAAkXn3tFIV+T6HQKOj9MZkmA==
      rbd_store_pool: nova
      rbd_store_user: nova
````

dirkmueller commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,75 @@
+            'cluster_network': list(local.cmd(master_minion, 'pillar.get', ['cluster_network']).items())[0][1]
+        },
+        'pools': {
+            'cinder': {
```
this name needs to be configurable (or at least prefixable) for the usecase of multiple openstack installations sharing a single storage cluster..

dirkmueller commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,75 @@
+                        list(local.cmd(master_minion, 'keyring.file', ['cinder']).items())[0][1]
+                    ]).items())[0][1]
+            },
+            'glance': {
```
same here

dirkmueller commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,75 @@
+                    ]).items())[0][1]
+            },
+            'cinder-backup': {},
+            'nova': {
```
same here

tserong commented:
@swiftgist, the above allows us to keep the same template keyring files and salt states no matter how many different OpenStack installs there are, it just results in a different set of keyring cache files being generated per prefix.

I'm not sure if the keycaps are correct at present, and I'm also not entirely clear on whether there actually needs to be a separate nova user & key, or if nova can just use the cinder credentials.

tserong commented:
(Mostly just a rebase, safe to ignore for the moment)

hemna commented:
For  openstack integration, there is no need for a nova keyring or user.  Nova uses the cinder keyring and user.   Nova just needs it's own pool and the cinder user needs access to rwx against the nova pool.

Also,
Here is the current config file that is being used by  SOC for consumption by the SES extension to enable SES integration.

```---

ses_cluster_configuration:
    ses_cluster_name: ceph
    ses_radosgw_url: "https://192.168.56.8:8080/swift/v1"

    conf_options:
        ses_fsid: d5d7c7cb-5858-3218-a36f-d028df7b1111
        ses_mon_initial_members: ses-osd2, ses-osd3, ses-osd1
        ses_mon_host: 192.168.56.8, 192.168.56.9, 192.168.56.7
        ses_public_network: 192.168.56.0/21
        ses_cluster_network: 192.168.56.0/21

    cinder:
        rbd_store_pool: cinder
        rbd_store_pool_user: cinder
        keyring_file_name: ceph.client.cinder.keyring

    cinder-backup:
        rbd_store_pool: backups
        rbd_store_pool_user: cinder_backup
        keyring_file_name: ceph.client.cinder-backup.keyring

    # Nova uses the cinder user to access the nova pool, cinder pool
    # So all we need here is the nova pool name.
    nova:
        rbd_store_pool: nova

    glance:
        rbd_store_pool: glance
        rbd_store_pool_user: glance
        keyring_file_name: ceph.client.glance.keyring

hemna commented:
If you notice in the config file, we also need the Rados gw URL for object storage support too.

tserong commented:
Thanks @hemna, working on it :-)

jschmid1 commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,212 @@
+    if you wish to have a single Ceph cluster provide storage for multiple
+    OpenStack deployments, use of the "prefix" parameter will alter the pool
+    names, for example specifying "prefix=cloud" will result in pools named
+    "cloud-images", "cloud-volumes", "cloud-backups" and "cloud-vms".
```
I think the prefix=cloud should be default here. That makes it quite clear in the first place. We still can make the prefix option available for let's say dc2-cloud-images (datacenter2).

tserong replied:
Good idea

jschmid1 commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,212 @@
+        section = "client." + rgw[0] + "." + rgw[1]
+        if not conf.has_section(section):
+            continue
+        if conf.has_option(section, "rgw frontends") and conf.has_option(section, "rgw dns name"):
```
A perfect candidate for a unit test :)

jschmid1 commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,212 @@
+    configure OpenStack.
+
+    This will create pools for use by glance, cinder, cinder-backup and nova.
+    By default, these pools will be named "images", "volumes", "backups" and
```
I think the prefix=cloud should be default here. That makes it quite clear in the first place. We still can make the prefix option available for let's say dc2-cloud-images (datacenter2).

jschmid1 commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,212 @@
+    __salt__ = salt.loader.minion_mods(__opts__, utils=__utils__)
+    master_minion = __salt__['master.minion']()
+
+    if "prefix" in kwargs:
```
Duplicate checks for `prefix in kwargs`

jschmid1 commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,212 @@
+    # Set up prefix for subsequent string concatenation to match what's done
+    # in the SLS files for keyring and pool names.
+    prefix = ""
+    if "prefix" in kwargs:
```
Duplicate checks for prefix in kwargs

jschmid1 commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,212 @@
+    if len(rgw_urls) == 1:
+        radosgw_url = rgw_urls[0]
+    elif len(rgw_urls) > 1:
+        # TODO: This is...  Not ideal.  But I don't have any better ideas right now.
```
You mean that you don't have a good way to distinct which rgw_url to return?

tserong replied:
Correct. That the string comes back as "FIXME: http://rgw1/ http://rgw2 ..." at least is easy for a human reading to know they need to make a choice, and is easy for an automation tool to match and fail on. If that's OK (or if we don't come up with anything better), then I'll remove the "TODO".

Hrm.  I wonder if I should maybe change it to just return a list of all RGW endpoints, rather than a weird string.

hemna replied:
I think making a list of all of the RGW endpoints would be better here.   The user can edit them if needed, or the integration in SOC will just pick the first in the list.

jschmid1 commented on srv/modules/runners/openstack.py:
```diff
@@ -0,0 +1,212 @@
+            'public_network': list(local.cmd(master_minion, 'pillar.get',
+                                             ['public_network']).items())[0][1],
+            'cluster_network': list(local.cmd(master_minion, 'pillar.get',
+                                              ['cluster_network']).items())[0][1]
```
I could see a lot of 'list index out of rage' errors when something before those calls fails. 

Maybe we should try to gather the pillar data before and validate it somehow. ( lots of somehows and maybes in there )

tserong replied:
:-)

jschmid1 commented on srv/salt/ceph/openstack/cinder-backup/key/default.sls:
```diff
@@ -0,0 +1,16 @@
+{% set prefix = pillar['openstack_prefix'] + "-" if 'openstack_prefix' in pillar else "" %}
```
Couldn't this be the standard:

`{% set prefix = salt['pillar.get']('openstack_prefix' + "-", "") %}`

tserong replied:
That would be rather less ugly. Thank you.

tserong replied:
Actually, that doesn't quite work. `salt['pillar.get']('openstack_prefix' + "-", "")` tries to get a value named 'openstack-prefix-'. I need to get a value named 'openstack-prefix', and append '-' to the value, iff the value is present. Would the following be clearer?

````
{% set prefix = salt['pillar.get']('openstack_prefix', "") %}
{% set prefix = prefix + "-" if prefix else "" %}
````

Or is there a neater one-liner than I'm just not seeing yet?

jschmid1 commented on srv/salt/ceph/openstack/cinder/auth/default.sls:
```diff
@@ -1,6 +1,7 @@
-{% set keyring_file = salt['keyring.file']('cinder') %}
+{% set prefix = pillar['openstack_prefix'] + "-" if 'openstack_prefix' in pillar else "" %}
```
https://github.com/SUSE/DeepSea/pull/1025/files#r196747950

jschmid1 commented on srv/salt/ceph/openstack/cinder/key/default.sls:
```diff
@@ -1,5 +1,5 @@
-
-{% set keyring_file = salt['keyring.file']('cinder') %}
+{% set prefix = pillar['openstack_prefix'] + "-" if 'openstack_prefix' in pillar else "" %}
```
https://github.com/SUSE/DeepSea/pull/1025/files#r196747950

jschmid1 commented on srv/salt/ceph/openstack/cinder/pool/default.sls:
```diff
@@ -1,7 +1,11 @@
-
-cinder pool:
+{% set prefix = pillar['openstack_prefix'] + "-" if 'openstack_prefix' in pillar else "" %}
```
https://github.com/SUSE/DeepSea/pull/1025/files#r196747950

jschmid1 commented on srv/salt/ceph/openstack/cinder-backup/pool/default.sls:
```diff
@@ -0,0 +1,11 @@
+{% set prefix = pillar['openstack_prefix'] + "-" if 'openstack_prefix' in pillar else "" %}
```
same comment as in https://github.com/SUSE/DeepSea/pull/1025/files#r196747950..

exemplary for all other occurrences

jschmid1 commented on srv/salt/ceph/smoketests/openstack/init.sls:
```diff
@@ -0,0 +1,88 @@
+# TODO:
+# - Verify the various users can access the appropriate pools and check
+#   ceph status, but can't do anything else.
+# - Delete the just-created pools and users, to leave the cluster somewhat
```
yes :)

jschmid1 commented on srv/salt/ceph/smoketests/openstack/init.sls:
```diff
@@ -0,0 +1,88 @@
+#   clean-ish?
+# - Add tests for the openstack.integrate runner itself, to make sure it
+#   returns correct data.
+# - Smoketests without prefix?  This would be destructive of any existing
```
That's what I was about to write. Adding a prefix takes a considerable ammount of different routes across the codebase. It might be wise to add a smoketest for the 'prefix route' aswell.

jschmid1 commented on srv/salt/ceph/smoketests/openstack/init.sls:
```diff
@@ -0,0 +1,88 @@
+#   ceph status, but can't do anything else.
+# - Delete the just-created pools and users, to leave the cluster somewhat
+#   clean-ish?
+# - Add tests for the openstack.integrate runner itself, to make sure it
```
as mentioned above :+1:

jschmid1 commented:
Oh, and a rebase atop master would be great. Pip is currently pulling a broken salt package. #1186

jschmid1 commented:
As I just ran into a `TypeError` as anticipated above, I'd generally recommend constructing strings with the `.format` method rather than using "a" + "b" notation.

hemna commented:
This looks like a better yaml that we can work with.   It won't take many changes on our side to consume this version of the config.

tserong commented:
- Rebased on master
- Pool names all start with "cloud-" by default
- Removed duplicate check for prefix in kwargs
- radosgw_urls (note: plural) is now a list (which can be an empty list, if there's no RGW instances)

(Still some more comments that need addressing, but that's all internal stuff like tests and more validation)

tserong commented:
That should be about it, now, except for the possibility of writing more tests :-)

swiftgist commented:
I took it for a spin from a fresh deployment and well, it just worked.  For those interested in output examples, here's a couple:

```
 # salt-run --out=yaml openstack.integrate
ceph_conf:
  cluster_network: 172.16.22.0/24
  fsid: 43ba3934-fbe9-354f-91be-f3f8e92e5092
  mon_host: 172.16.21.13, 172.16.21.11, 172.16.21.12
  mon_initial_members: mon1, mon2, mon3
  public_network: 172.16.21.0/24
cinder:
  key: AQB7dTpbAAAAABAAbRhkuYwNpHFefrwDOn27Tw==
  rbd_store_pool: cloud-volumes
  rbd_store_user: cinder
cinder-backup:
  key: AQB7dTpbAAAAABAAOFskzWdSglMKbh+5ULQWDw==
  rbd_store_pool: cloud-backups
  rbd_store_user: cinder-backup
glance:
  key: AQB7dTpbAAAAABAAZDaKdmczwZXwijhAnloesg==
  rbd_store_pool: cloud-images
  rbd_store_user: glance
nova:
  rbd_store_pool: cloud-vms
radosgw_urls:
- http://rgw2.ceph:80/swift/v1
- http://rgw1.ceph:80/swift/v1
- http://rgw3.ceph:80/swift/v1

# salt-run --out=yaml openstack.integrate prefix=teamA
ceph_conf:
  cluster_network: 172.16.22.0/24
  fsid: 43ba3934-fbe9-354f-91be-f3f8e92e5092
  mon_host: 172.16.21.12, 172.16.21.13, 172.16.21.11
  mon_initial_members: mon2, mon1, mon3
  public_network: 172.16.21.0/24
cinder:
  key: AQBrdjpbAAAAABAARsDenY8WFexmYoehyq0/rg==
  rbd_store_pool: teamA-cloud-volumes
  rbd_store_user: teamA-cinder
cinder-backup:
  key: AQBrdjpbAAAAABAAcN6tlVeG/0K16cvotdiAFQ==
  rbd_store_pool: teamA-cloud-backups
  rbd_store_user: teamA-cinder-backup
glance:
  key: AQBrdjpbAAAAABAAUbeAwjhJ4DRwJEmAxQU9Zg==
  rbd_store_pool: teamA-cloud-images
  rbd_store_user: teamA-glance
nova:
  rbd_store_pool: teamA-cloud-vms
radosgw_urls:
- http://rgw1.ceph:80/swift/v1
- http://rgw3.ceph:80/swift/v1
- http://rgw2.ceph:80/swift/v1
```

Smoketest ran fine as well and cleans up after itself. :)

***
The only observations that I have is that I do not know what `cinder-backup` is.  I can guess and I'll probably be right, but a README wouldn't hurt the unfortunate admins learning Salt, OpenStack and Ceph simultaneously.  I am only thinking of `cinder-backup` here since it does not completely match an OpenStack component.

The runner doesn't have a `help` function.

This last one can go in a different PR and not hold this one up.  What do you think of creating a master module for `openstack.prefix`?  I'm thinking that drops a Jinja line from nearly every state file.  

***
These are just out loud thoughts that I suspect has already been discussed and I'm just behind. :)  Will `openstack_prefix` be multi-valued at any point?  Or should the prefix stay a kwarg and not be a pillar variable?  (Trying to think of the benefits either way.)  Should `openstack_prefix` be moved to `ceph:openstack_prefix` to prevent name collisions?

kshtsk commented:
retest this please

tserong commented:
> What do you think of creating a master module for openstack.prefix? I'm thinking that drops a Jinja line from nearly every state file.

I hadn't thought of that, but I like the idea of anything that drops irritating Jinja :-)

> Will openstack_prefix be multi-valued at any point?

I don't expect it to be. The original idea was "give me some pools and users", or, "give me some pools and users suitable for a second (or third, or fourth, or, ...) openstack instance", so each invocation could nominate a prefix or not.

tserong commented:
Does anyone have any objection to merging this thing as-is? @jschmid1? Should I request review from anyone else?

kshtsk commented:
retest this please

kshtsk commented:
@jschmid1 @tserong the linter is failing in saltpylint, this is reproducible on master branch as well

tserong commented:
If that lint failure is what I think it is, it should be fixed in master already by eb3a71be

tserong commented:
Rebased (I can still squash some of these commits if anyone wants me to)

tserong commented:
retest this please

tserong commented:
Ah! I wasn't seeing the lint failure earlier in my local checkout, but after `rm -rf .tox ; tox -e lint`, I now get:

````
GLOB sdist-make: /home/tserong/src/github/SUSE/DeepSea/setup.py
lint create: /home/tserong/src/github/SUSE/DeepSea/.tox/lint
lint installdeps: ipaddress, salt<2017.0.0, saltpylint
lint inst: /home/tserong/src/github/SUSE/DeepSea/.tox/dist/deepsea-0.9.5+75.g8c703c1d.zip
lint installed: astroid==2.0,certifi==2018.4.16,chardet==3.0.4,deepsea==0.9.5+75.g8c703c1d,idna==2.7,ipaddress==1.0.22,isort==4.3.4,Jinja2==2.10,lazy-object-proxy==1.3.1,MarkupSafe==1.0,mccabe==0.6.1,modernize==0.5,msgpack-python==0.5.6,pep8==1.6.2,pycrypto==2.6.1,pylint==2.0.0,PyYAML==3.13,pyzmq==17.1.0,requests==2.19.1,salt==2016.11.9,SaltPyLint==2017.12.15,six==1.11.0,tornado==5.1,typed-ast==1.1.0,urllib3==1.23,wrapt==1.10.11
lint runtests: PYTHONHASHSEED='1872628738'
lint runtests: commands[0] | pylint --rcfile=.pylintrc srv/
Traceback (most recent call last):
  File "/home/tserong/src/github/SUSE/DeepSea/.tox/lint/bin/pylint", line 11, in <module>
    sys.exit(run_pylint())
  File "/home/tserong/src/github/SUSE/DeepSea/.tox/lint/lib/python3.6/site-packages/pylint/__init__.py", line 19, in run_pylint
    Run(sys.argv[1:])
  File "/home/tserong/src/github/SUSE/DeepSea/.tox/lint/lib/python3.6/site-packages/pylint/lint.py", line 1359, in __init__
    linter.load_plugin_modules(plugins)
  File "/home/tserong/src/github/SUSE/DeepSea/.tox/lint/lib/python3.6/site-packages/pylint/lint.py", line 520, in load_plugin_modules
    module.register(self)
  File "/home/tserong/src/github/SUSE/DeepSea/.tox/lint/lib/python3.6/site-packages/saltpylint/smartup.py", line 43, in register
    MANAGER.register_transform(nodes.Class, rootlogger_transform)
AttributeError: module 'astroid.nodes' has no attribute 'Class'
ERROR: InvocationError: '/home/tserong/src/github/SUSE/DeepSea/.tox/lint/bin/pylint --rcfile=.pylintrc srv/'
````

That looks to me like maybe a bug in astroid (whatever astroid is), rather than a fault in DeepSea.

kshtsk commented:
@tserong Rebase is a good step, but 'master' has this issue as well.
Try to reproduce it on your local host in clean environment: 
```git clone https://github.com/suse/deepsea && cd deepsea && virtualenv -p python2.7 v && . v/bin/activate && pip install --upgrade pip && pip install tox && make lint```

kshtsk commented:
@tserong Interesting that on Mac I have the error message you have, but on Leap 42.3 the error is: ```AttributeError: 'module' object has no attribute 'Class```

tserong commented:
I really think this is somehow a bug in saltpylint, or one of its dependencies (or some combination of the two), and this warrants a separate issue.

kshtsk commented:
@tserong me too

kshtsk commented:
so, you guys need to fix saltpylint issue in separate PR, merge it, and then go with this PR after rebase,
or... agree that you do not run `make lint` or ignore it

jschmid1 commented:
@tserong I'd like to see the commits squashed/grouped for backporting reasons.

I experienced it to be the easiest if we separate unittests, smoketests and the implementation.

tserong commented:
Done, all those cleanup commits are squashed.

jschmid1 commented:
:+1: from my side @kbaikov did you

jschmid1 commented:
:+1: from my side. @kbaikov Did you have time to give this a test also?

kshtsk commented:
@susebot run teuthology

kshtsk commented:
@susebot run teuthology

jschmid1 commented:
@susebot run teuthology

jschmid1 commented:
The failed run in teuthology look like 'dead' jobs.

smithfarm commented:
I looked at the health-ok teuthology failure - it happened because teuthology failed to start the target VM.

kshtsk commented:
@smithfarm Interesting that teuthology fails with

```
MaxWhileTries: 'get ip 36aa0852-1cce-47e3-a1e3-ede3afe06056' reached maximum tries (30) after waiting for 60 seconds
```

And the reason that vm has status ERROR with message:
 _Build of instance 36aa0852-1cce-47e3-a1e3-ede3afe06056 aborted: Failed to allocate the network(s), not rescheduling._

That means, we can try and improve it to detect ERROR status earlier and try to recreate the VM.

kbaikov commented:
This looks good to me, but i would like to test it with SOC7 and sles15sp3.
@tserong Could you backport it, please?

tserong commented:
> @tserong Could you backport it, please?

Yep.

@jschmid1, I assume the thing to do is create another wip-* branch off the SES5 branch, then cherry pick (and fixup as necessary) to there, then finally open a PR against the SES5 branch?

jschmid1 commented:
> @jschmid1, I assume the thing to do is create another wip-* branch off the SES5 branch, then cherry pick (and fixup as necessary) to there, then finally open a PR against the SES5 branch?

perfectly right. There is also a wiki entry: https://github.com/SUSE/DeepSea/wiki/Backporting

tserong commented:
Rebased
